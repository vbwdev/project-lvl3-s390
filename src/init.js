import 'bootstrap';
import 'bootstrap/dist/css/bootstrap.min.css';

import axios from 'axios';
import isUrl from 'validator/lib/isURL';
import WatchJS from 'melanke-watchjs';
import { isEmpty } from 'lodash';

import getAlertHtml from './components/alert';
import getArticlesNodes from './components/articlesList';
import getChannelsListHtml from './components/channelsList';
import parseRssFeed from './parser';

const { watch } = WatchJS;

const corsProxyUrl = 'https://cors-anywhere.herokuapp.com/';
const getWithProxy = url => axios.get(`${corsProxyUrl}${url}`);

const getUrlFormState = (stateName, oldState, { url, error } = {}) => {
  switch (stateName) {
    case 'validation-error':
      return {
        canSubscribe: false,
        formError: null,
        hasUrlError: true,
        isFetching: false,
        urlInputValue: url || oldState.addUrlForm.urlInputValue,
      };
    case 'fetching':
      return {
        canSubscribe: false,
        formError: null,
        hasUrlError: false,
        isFetching: true,
        urlInputValue: url || oldState.addUrlForm.urlInputValue,
      };
    case 'fetching-error':
      return {
        canSubscribe: true,
        formError: error,
        hasUrlError: false,
        isFetching: false,
        urlInputValue: url || oldState.addUrlForm.urlInputValue,
      };
    case 'validation-no-error':
      return {
        canSubscribe: true,
        formError: null,
        hasUrlError: false,
        isFetching: false,
        urlInputValue: url || oldState.addUrlForm.urlInputValue,
      };
    case 'empty':
      return {
        canSubscribe: false,
        formError: null,
        hasUrlError: false,
        isFetching: false,
        urlInputValue: '',
      };
    default:
      console.error(`Unknown state name "${stateName}"`);
      return oldState.addUrlForm;
  }
};

export default () => {
  const state = {
    addUrlForm: {
      ...getUrlFormState('empty'),
    },
    rssUrls: [],
    articles: {},
    channels: [],
    articleDescriptionModal: {
      description: null,
      link: null,
      title: null,
    },
    isBackgroundRssFeedsUpdating: false,
  };

  const setUrlFormState = (stateName, values) => {
    Object.assign(state, { addUrlForm: getUrlFormState(stateName, state, values) });
  };

  const addRssFeed = url =>
    getWithProxy(url)
      .then(({ data }) => parseRssFeed(data))
      .then(({ articles, channel }) => {
        state.rssUrls = [...state.rssUrls, url];
        state.articles = {
          ...state.articles,
          ...articles,
        };
        state.channels = [...state.channels, channel];
      });

  const updateRssFeedArticles = url =>
    getWithProxy(url)
      .then(({ data }) => parseRssFeed(data))
      .then(({ articles }) => {
        state.articles = {
          ...articles,
          ...state.articles,
        };
      });

  const updateRssFeedsArticles = () => {
    if (isEmpty(state.rssUrls)) {
      return Promise.resolve();
    }

    const promises = state.rssUrls.map(updateRssFeedArticles);
    return Promise.all(promises);
  };

  const getChannelsList = () => state.channels;

  const getArticlesList = () => Object.values(state.articles);

  const isDuplicatedUrl = url => state.rssUrls.includes(url);

  const initBackgroundRssFeedsUpdating = () => {
    setTimeout(() => {
      updateRssFeedsArticles().finally(initBackgroundRssFeedsUpdating);
    }, 5000);
  };

  const updateModalContent = ({ description, link, title }) => {
    state.articleDescriptionModal = {
      description,
      link,
      title,
    };
  };

  const rssUrlForm = document.querySelector('.js-rss-url-form');
  const rssUrlInput = document.querySelector('.js-rss-url-input');
  const rssUrlSubmitButton = document.querySelector('.js-rss-url-submit-button');
  const formAlerts = document.querySelector('.js-form-alerts');
  const channelsList = document.querySelector('.js-rss-channels-list');
  const articlesList = document.querySelector('.js-rss-articles-list');
  const articleDescriptionModal = document.querySelector('#articleDescriptionModal');

  rssUrlInput.addEventListener('input', e => {
    if (state.addUrlForm.isFetching) {
      return false;
    }

    const { value: url } = e.target;
    if (url === '') {
      setUrlFormState('empty');
    } else if (!isUrl(url) || isDuplicatedUrl(url)) {
      setUrlFormState('validation-error', { url });
    } else {
      setUrlFormState('validation-no-error', { url });
    }
    return true;
  });

  rssUrlForm.addEventListener('submit', e => {
    e.preventDefault();

    if (!state.addUrlForm.canSubscribe || state.addUrlForm.isFetching) {
      return false;
    }

    setUrlFormState('fetching');
    const rssFeedUrl = rssUrlInput.value;
    return addRssFeed(rssFeedUrl)
      .then(() => {
        setUrlFormState('empty');
        if (!state.isBackgroundRssFeedsUpdating) {
          initBackgroundRssFeedsUpdating();
          state.isBackgroundRssFeedsUpdating = true;
        }
      })
      .catch(error => {
        setUrlFormState('fetching-error', { error });
      });
  });

  watch(state, 'addUrlForm', () => {
    const { canSubscribe, formError, hasUrlError, isFetching, urlInputValue } = state.addUrlForm;
    rssUrlInput.value = urlInputValue;
    rssUrlInput.classList.toggle('is-invalid', hasUrlError);
    rssUrlSubmitButton.disabled = !canSubscribe;
    rssUrlSubmitButton.innerHTML = isFetching
      ? '<span class="spinner-border spinner-border-sm" role="status" aria-hidden="true"></span>'
      : 'Submit';
    formAlerts.innerHTML = formError ? getAlertHtml(formError) : '';
  });

  watch(state, 'channels', () => {
    const channels = getChannelsList();
    channelsList.innerHTML = getChannelsListHtml(channels);
  });

  watch(state, 'articles', () => {
    const articles = getArticlesList();
    const articlesNodes = getArticlesNodes(articles, updateModalContent);
    articlesList.innerHTML = '';
    articlesList.append(...articlesNodes);
  });

  watch(state, 'articleDescriptionModal', () => {
    const { description, link, title } = state.articleDescriptionModal;
    articleDescriptionModal.querySelector('.js-modal-description').textContent = description;
    articleDescriptionModal.querySelector('.js-modal-link').href = link;
    articleDescriptionModal.querySelector('.js-modal-title').textContent = title;
  });
};
