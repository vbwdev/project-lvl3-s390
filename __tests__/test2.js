/**
 * 'should update articles by timeout' â€“ is long test and it broke other tests.
 * TODO Rework long test
 */

import fs from 'fs';
import path from 'path';
import { promisify } from 'util';
import { html } from 'js-beautify';
import keycode from 'keycode';
import nock from 'nock';

import init from '../src/init';

const htmlOptions = {
  indent_size: 2,
  preserve_newlines: false,
};

const pressKey = (key, el = document.body, value = key) => {
  const keyCode = keycode(key);
  const e = new KeyboardEvent('input', { keyCode });
  el.value = value; // eslint-disable-line no-param-reassign
  el.setAttribute('value', value);
  el.dispatchEvent(e);
};

const readFile = promisify(fs.readFile);

const proxyUrl = 'https://cors-anywhere.herokuapp.com/';
const proxyHeaders = { 'access-control-allow-origin': '*' };

const articlesListSelector = '.js-rss-articles-list';
const channelsListSelector = '.js-rss-channels-list';

describe('rss reader', () => {
  nock.disableNetConnect();
  const rssFeedHexletPart1 = fs.readFileSync(
    `${__dirname}/__fixtures__/rss-feed-hexlet-part-1.txt`,
    'utf8',
  );
  const rssFeedHexletPart2 = fs.readFileSync(
    `${__dirname}/__fixtures__/rss-feed-hexlet-part-2.txt`,
    'utf8',
  );
  let rssUrlForm;
  let rssUrlInput;

  beforeEach(async () => {
    const pathToHtml = path.resolve(__dirname, '../template.html');
    document.body.innerHTML = await readFile(pathToHtml, 'utf8');
    init();
    rssUrlInput = document.querySelector('.js-rss-url-input');
    rssUrlForm = document.querySelector('.js-rss-url-form');
  });

  test('should update articles by timeout', done => {
    jest.setTimeout(10000);

    const url = 'https://good-url.com';
    nock(proxyUrl)
      .defaultReplyHeaders(proxyHeaders)
      .get(`/${url}`)
      .reply(200, rssFeedHexletPart1)
      .defaultReplyHeaders(proxyHeaders)
      .get(`/${url}`)
      .reply(200, rssFeedHexletPart2);

    pressKey('m', rssUrlInput, url);
    rssUrlForm.dispatchEvent(new Event('submit'));

    setTimeout(() => {
      expect(
        html(document.querySelector(channelsListSelector).innerHTML, htmlOptions),
      ).toMatchSnapshot();
      expect(
        html(document.querySelector(articlesListSelector).innerHTML, htmlOptions),
      ).toMatchSnapshot();
    }, 100);

    setTimeout(() => {
      expect(
        html(document.querySelector(channelsListSelector).innerHTML, htmlOptions),
      ).toMatchSnapshot();
      expect(
        html(document.querySelector(articlesListSelector).innerHTML, htmlOptions),
      ).toMatchSnapshot();
      done();
    }, 7000);
  });
});
