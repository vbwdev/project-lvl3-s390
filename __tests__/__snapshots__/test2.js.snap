// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rss reader should update articles by timeout 1`] = `
"<li class=\\"list-group-item\\">
  <h4>Новые уроки на Хекслете</h4>Практические уроки по программированию
</li>"
`;

exports[`rss reader should update articles by timeout 2`] = `
"<div class=\\"list-group-item\\">
  <h4><a target=\\"_blank\\" href=\\"https://ru.hexlet.io/courses/main/lessons/functional-contest-03-2015/theory_unit\\">Ареалы проживания членистоногих. Конкурс по функциональному программированию 03.2015. / Главные испытания</a></h4><button type=\\"button\\" class=\\"btn btn-primary js-show-article-modal-button\\" data-target=\\"#articleDescriptionModal\\" data-toggle=\\"modal\\">More</button>
</div>
<div class=\\"list-group-item\\">
  <h4><a target=\\"_blank\\" href=\\"https://ru.hexlet.io/courses/frameworks/lessons/rails_getting_started/theory_unit\\">Rails для начинающих / Фреймворки</a></h4><button type=\\"button\\" class=\\"btn btn-primary js-show-article-modal-button\\" data-target=\\"#articleDescriptionModal\\" data-toggle=\\"modal\\">More</button>
</div>
<div class=\\"list-group-item\\">
  <h4><a target=\\"_blank\\" href=\\"https://ru.hexlet.io/courses/main/lessons/pipeline/theory_unit\\">Pipeline challenge / Главные испытания</a></h4><button type=\\"button\\" class=\\"btn btn-primary js-show-article-modal-button\\" data-target=\\"#articleDescriptionModal\\" data-toggle=\\"modal\\">More</button>
</div>
<div class=\\"list-group-item\\">
  <h4><a target=\\"_blank\\" href=\\"https://ru.hexlet.io/courses/js-setup-environment/lessons/npx/theory_unit\\">NPX / JS: Настройка окружения</a></h4><button type=\\"button\\" class=\\"btn btn-primary js-show-article-modal-button\\" data-target=\\"#articleDescriptionModal\\" data-toggle=\\"modal\\">More</button>
</div>
<div class=\\"list-group-item\\">
  <h4><a target=\\"_blank\\" href=\\"https://ru.hexlet.io/courses/js-setup-environment/lessons/dependencies/theory_unit\\">Зависимости / JS: Настройка окружения</a></h4><button type=\\"button\\" class=\\"btn btn-primary js-show-article-modal-button\\" data-target=\\"#articleDescriptionModal\\" data-toggle=\\"modal\\">More</button>
</div>
<div class=\\"list-group-item\\">
  <h4><a target=\\"_blank\\" href=\\"https://ru.hexlet.io/courses/js-setup-environment/lessons/publish/theory_unit\\">Публикация пакета / JS: Настройка окружения</a></h4><button type=\\"button\\" class=\\"btn btn-primary js-show-article-modal-button\\" data-target=\\"#articleDescriptionModal\\" data-toggle=\\"modal\\">More</button>
</div>
<div class=\\"list-group-item\\">
  <h4><a target=\\"_blank\\" href=\\"https://ru.hexlet.io/courses/js-asynchronous-programming/lessons/async-await/theory_unit\\">Async/Await / JS: Асинхронное программирование</a></h4><button type=\\"button\\" class=\\"btn btn-primary js-show-article-modal-button\\" data-target=\\"#articleDescriptionModal\\" data-toggle=\\"modal\\">More</button>
</div>
<div class=\\"list-group-item\\">
  <h4><a target=\\"_blank\\" href=\\"https://ru.hexlet.io/courses/js-asynchronous-programming/lessons/new-promise/theory_unit\\">new Promise / JS: Асинхронное программирование</a></h4><button type=\\"button\\" class=\\"btn btn-primary js-show-article-modal-button\\" data-target=\\"#articleDescriptionModal\\" data-toggle=\\"modal\\">More</button>
</div>
<div class=\\"list-group-item\\">
  <h4><a target=\\"_blank\\" href=\\"https://ru.hexlet.io/courses/js-asynchronous-programming/lessons/promises-catch/theory_unit\\">Обработка ошибок в промисах / JS: Асинхронное программирование</a></h4><button type=\\"button\\" class=\\"btn btn-primary js-show-article-modal-button\\" data-target=\\"#articleDescriptionModal\\" data-toggle=\\"modal\\">More</button>
</div>
<div class=\\"list-group-item\\">
  <h4><a target=\\"_blank\\" href=\\"https://ru.hexlet.io/courses/js-asynchronous-programming/lessons/promise-all/theory_unit\\">Promise.all / JS: Асинхронное программирование</a></h4><button type=\\"button\\" class=\\"btn btn-primary js-show-article-modal-button\\" data-target=\\"#articleDescriptionModal\\" data-toggle=\\"modal\\">More</button>
</div>"
`;

exports[`rss reader should update articles by timeout 3`] = `
"<li class=\\"list-group-item\\">
  <h4>Новые уроки на Хекслете</h4>Практические уроки по программированию
</li>"
`;

exports[`rss reader should update articles by timeout 4`] = `
"<div class=\\"list-group-item\\">
  <h4><a target=\\"_blank\\" href=\\"https://ru.hexlet.io/courses/main/lessons/physics\\">Почему эксперментально верифицируема суспензия?</a></h4><button type=\\"button\\" class=\\"btn btn-primary js-show-article-modal-button\\" data-target=\\"#articleDescriptionModal\\" data-toggle=\\"modal\\">More</button>
</div>
<div class=\\"list-group-item\\">
  <h4><a target=\\"_blank\\" href=\\"https://ru.hexlet.io/courses/frameworks/lessons/rails_unit\\">Ультрафиолетовый взрыв: зеркало или темная материя?</a></h4><button type=\\"button\\" class=\\"btn btn-primary js-show-article-modal-button\\" data-target=\\"#articleDescriptionModal\\" data-toggle=\\"modal\\">More</button>
</div>
<div class=\\"list-group-item\\">
  <h4><a target=\\"_blank\\" href=\\"https://ru.hexlet.io/courses/main/lessons/functional-contest-03-2015/theory_unit\\">Ареалы проживания членистоногих. Конкурс по функциональному программированию 03.2015. / Главные испытания</a></h4><button type=\\"button\\" class=\\"btn btn-primary js-show-article-modal-button\\" data-target=\\"#articleDescriptionModal\\" data-toggle=\\"modal\\">More</button>
</div>
<div class=\\"list-group-item\\">
  <h4><a target=\\"_blank\\" href=\\"https://ru.hexlet.io/courses/frameworks/lessons/rails_getting_started/theory_unit\\">Rails для начинающих / Фреймворки</a></h4><button type=\\"button\\" class=\\"btn btn-primary js-show-article-modal-button\\" data-target=\\"#articleDescriptionModal\\" data-toggle=\\"modal\\">More</button>
</div>
<div class=\\"list-group-item\\">
  <h4><a target=\\"_blank\\" href=\\"https://ru.hexlet.io/courses/main/lessons/pipeline/theory_unit\\">Pipeline challenge / Главные испытания</a></h4><button type=\\"button\\" class=\\"btn btn-primary js-show-article-modal-button\\" data-target=\\"#articleDescriptionModal\\" data-toggle=\\"modal\\">More</button>
</div>
<div class=\\"list-group-item\\">
  <h4><a target=\\"_blank\\" href=\\"https://ru.hexlet.io/courses/js-setup-environment/lessons/npx/theory_unit\\">NPX / JS: Настройка окружения</a></h4><button type=\\"button\\" class=\\"btn btn-primary js-show-article-modal-button\\" data-target=\\"#articleDescriptionModal\\" data-toggle=\\"modal\\">More</button>
</div>
<div class=\\"list-group-item\\">
  <h4><a target=\\"_blank\\" href=\\"https://ru.hexlet.io/courses/js-setup-environment/lessons/dependencies/theory_unit\\">Зависимости / JS: Настройка окружения</a></h4><button type=\\"button\\" class=\\"btn btn-primary js-show-article-modal-button\\" data-target=\\"#articleDescriptionModal\\" data-toggle=\\"modal\\">More</button>
</div>
<div class=\\"list-group-item\\">
  <h4><a target=\\"_blank\\" href=\\"https://ru.hexlet.io/courses/js-setup-environment/lessons/publish/theory_unit\\">Публикация пакета / JS: Настройка окружения</a></h4><button type=\\"button\\" class=\\"btn btn-primary js-show-article-modal-button\\" data-target=\\"#articleDescriptionModal\\" data-toggle=\\"modal\\">More</button>
</div>
<div class=\\"list-group-item\\">
  <h4><a target=\\"_blank\\" href=\\"https://ru.hexlet.io/courses/js-asynchronous-programming/lessons/async-await/theory_unit\\">Async/Await / JS: Асинхронное программирование</a></h4><button type=\\"button\\" class=\\"btn btn-primary js-show-article-modal-button\\" data-target=\\"#articleDescriptionModal\\" data-toggle=\\"modal\\">More</button>
</div>
<div class=\\"list-group-item\\">
  <h4><a target=\\"_blank\\" href=\\"https://ru.hexlet.io/courses/js-asynchronous-programming/lessons/new-promise/theory_unit\\">new Promise / JS: Асинхронное программирование</a></h4><button type=\\"button\\" class=\\"btn btn-primary js-show-article-modal-button\\" data-target=\\"#articleDescriptionModal\\" data-toggle=\\"modal\\">More</button>
</div>
<div class=\\"list-group-item\\">
  <h4><a target=\\"_blank\\" href=\\"https://ru.hexlet.io/courses/js-asynchronous-programming/lessons/promises-catch/theory_unit\\">Обработка ошибок в промисах / JS: Асинхронное программирование</a></h4><button type=\\"button\\" class=\\"btn btn-primary js-show-article-modal-button\\" data-target=\\"#articleDescriptionModal\\" data-toggle=\\"modal\\">More</button>
</div>
<div class=\\"list-group-item\\">
  <h4><a target=\\"_blank\\" href=\\"https://ru.hexlet.io/courses/js-asynchronous-programming/lessons/promise-all/theory_unit\\">Promise.all / JS: Асинхронное программирование</a></h4><button type=\\"button\\" class=\\"btn btn-primary js-show-article-modal-button\\" data-target=\\"#articleDescriptionModal\\" data-toggle=\\"modal\\">More</button>
</div>"
`;
